File: 443-string-compression/string-compression.py
class Solution(object):
    def compress(self, chars):
        """
        :type chars: List[str]
        :rtype: int
        """
        i = 0
        char = chars[0]
        dupes = 0

        while i < len(chars) - 1:
            # print("i: {}, char: {}, dupes: {}, chars: {}".format(i, char, dupes, chars))
            
            if chars[i] != chars[i + 1]:
                if dupes > 0:
                    # compress if dupes > 0
                    for digit in str(dupes + 1)[::-1]:
                        chars.insert(i + 1, digit)
                    i += len(str(dupes + 1)) + 1
                else:
                    i += 1
                    
                dupes = 0
                char = chars[i]

            else:
                dupes += 1
                chars.pop(i + 1)

        if dupes > 0:
            for digit in str(dupes + 1)[::-1]:
                chars.insert(i + 1, digit)

        return len(chars)

File: 443-string-compression/string-compression.java
class Solution {
    public int compress(char[] chars) {
        int left = 0;
        int right = 0;
        int write_ptr = 0;

        while (right < chars.length) {
            if (chars[left] == chars[right]) {
                right++;
            } else {
                // write current consecutive character to the array then increment write_ptr
                chars[write_ptr++] = chars[left];

                // write count of current consecutive character to the array
                if ((right - left) > 1) {
                    for (char digit : Integer.toString(right - left).toCharArray()) {
                        chars[write_ptr++] = digit;
                    }
                }
                left = right;
            }
        }

        // perform last compression if necessary
        chars[write_ptr++] = chars[left];
        if ((right - left) > 1) {
            for (char digit : Integer.toString(right - left).toCharArray()) {
                chars[write_ptr++] = digit;
            }
        }

        //System.out.println(Arrays.toString(chars));

        // return length of new array
        return write_ptr;
    }
}